Content.Server/GameObjects/Components/Chemistry/MovespeedStatusComponent.cs

```diff
@@ -0,0 +1,61 @@
+using Content.Shared.Movement.Components;
+using Robust.Shared.GameObjects;
+using Robust.Shared.ViewVariables;
+using System;
+using System.Threading;
+
+namespace Content.Server.GameObjects.Components.Chemistry
+{
+    //TODO: refactor movement modifier component because this is a pretty poor solution
+    [RegisterComponent]
+    class MovespeedStatusComponent : Component, IMoveSpeedModifier
+    {
+        
+        public override string Name => "ChemicalMovementSpeedChangeStatus";
+
+        [ViewVariables]
+        private float walkSpeedModifier;
+        public float WalkSpeedModifier
+        {
+            get => walkSpeedModifier;
+            set
+            {
+                walkSpeedModifier = value;
+                ResetTimer();
+            }
+        }
+        [ViewVariables]
+        private float sprintSpeedModifier;
+        public float SprintSpeedModifier
+        {
+            get => sprintSpeedModifier;
+            set
+            {
+                sprintSpeedModifier = value;
+                ResetTimer();
+            }
+        }
+
+        public int EffectTime { get; set; }
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            Owner.SpawnTimer(EffectTime, ResetModifiers, _cancellation.Token);
+        }
+
+        private void ResetModifiers()
+        {
+            WalkSpeedModifier = 0;
+            SprintSpeedModifier = 0;
+        }
+
+        private void ResetTimer()
+        {
+            _cancellation.Cancel();
+            Owner.SpawnTimer(EffectTime, ResetModifiers, _cancellation.Token);
+        }
```

Comment:
You shouldn't re-use the token, you should make a new one every time.

Make CancellationTokenSource nullable then do the following:
```suggestion
        private readonly CancellationTokenSource? _cancellation;

        private void ResetModifiers()
        {
            _cancellation?.Cancel();
            _walkSpeedModifier = 0;
            _sprintSpeedModifier = 0;
        }

        private void ResetTimer()
        {
            _cancellation?.Cancel();
            _cancellation = new CancellationTokenSource();
            Owner.SpawnTimer(EffectTime, ResetModifiers, _cancellation.Token);
        }
```

If you really wanted to minmax this you could have ResetModifiers nuke the component itself.